subprojects/core/src/main/java/org/gradle/api/internal/CompositeDomainObjectSet.java

```diff
@@ -149,6 +149,9 @@ public void all(Action<? super T> action) {
         private final List<DomainObjectCollection<? extends T>> store = Lists.newLinkedList();
 
         public boolean containsCollection(DomainObjectCollection<? extends T> collection) {
+            if (store.isEmpty()) {
```

Comment:
Actually, I did some more experiments and the `Iterators` were only allocated because I was profiling. When run in "production mode", escape analysis optimizes them away. So I think this additional code can go away.